src/cloud-api-adaptor/pkg/securecomms/ppssh/ppsecrets.go

```diff
@@ -14,44 +13,54 @@ type GetSecret func(name string) ([]byte, error)
 
 func NewPpSecrets(getSecret GetSecret) *PpSecrets {
 	return &PpSecrets{
-		keys:      []string{},
 		secrets:   make(map[string][]byte),
 		getSecret: getSecret,
 	}
 }
 
-func (fs *PpSecrets) AddKey(key string) {
-	fs.keys = append(fs.keys, key)
+func (sec *PpSecrets) AddKey(key string) {
+	if _, ok := sec.secrets[key]; ok {
+		return
+	}
+	sec.secrets[key] = nil
 }
 
-func (fs *PpSecrets) GetKey(key string) []byte {
-	return fs.secrets[key]
+func (sec *PpSecrets) GetKey(key string) []byte {
+	return sec.secrets[key]
 }
 
-func (fs *PpSecrets) Go() {
-	sleeptime := time.Duration(1)
+func (sec *PpSecrets) SetKey(key string, keydata []byte) {
+	sec.secrets[key] = keydata
+}
 
-	for len(fs.keys) > 0 {
-		key := fs.keys[0]
-		logger.Printf("PpSecrets obtaining key %s", key)
+func (sec *PpSecrets) Go() {
+	sleeptime := time.Duration(1)
 
-		data, err := fs.getSecret(key)
-		if err == nil && len(data) > 0 {
-			logger.Printf("PpSecrets %s success", key)
-			fs.secrets[key] = data
-			fs.keys = fs.keys[1:]
+	for key, keydata := range sec.secrets {
+		if keydata != nil {
 			continue
 		}
-		if err != nil {
-			logger.Printf("PpSecrets %s getSecret err: %v", key, err)
-		} else {
-			logger.Printf("PpSecrets %s getSecret returned an empty secret", key)
-		}
+		logger.Printf("PpSecrets obtaining key %s", key)
```

Comment:
How are the WN and PP keys  derived when Trustee is not there ?

Reply:
when Trustee is not there, WN and PP keys are sent using deamonConfig and PpSecrets.SetKey() is used to set them into the PpSecrets object.

PpSecrets.Go() is never called in this case - so this code is only used when attestation is used.

Reply:
Where is the WN and PP key generated before sending via daemonconfig ?

Reply:
See func (s *cloudService) CreateVM() in cloud.go. 
This is where `daemonConfig.PpPrivateKey` and `daemonConfig.WnPublicKey` are set. 
PpPrivateKey comes from s.sshClient.InitPP() which is executed once per PP and where the private PP key is generated. 
WnPublicKey comes from s.sshClient.GetWnPublicKey() - originally generated once per cluster in InitSshClient()